packages/core/src/modules/proofs/formats/presentation-exchange/PresentationExchangeProofFormatService.ts

```diff
@@ -0,0 +1,309 @@
+      throw new AriesFrameworkError('Invalid amount of verifiable presentations. Only one is allowed.')
+    }
+
+    const data = {
+      presentation_submission: presentation.presentationSubmission,
+      ...presentation.verifiablePresentations[0],
+    }
+
+    const attachment = this.getFormatData(data, format.attachmentId)
+
+    return { attachment, format }
+  }
+
+  public async processPresentation(
+    agentContext: AgentContext,
+    { requestAttachment, attachment }: ProofFormatProcessPresentationOptions
+  ): Promise<boolean> {
+    const ps = this.presentationExchangeService(agentContext)
+    const { presentation_definition: presentationDefinition } = requestAttachment.getDataAsJson<{
+      presentation_definition: PresentationDefinition
+    }>()
+    const presentation = attachment.getDataAsJson<
+      W3cVerifiablePresentation & { presentation_submission: PresentationSubmission }
+    >()
+
+    try {
+      ps.validatePresentationDefinition(presentationDefinition)
+      if (presentation.presentation_submission) {
+        ps.validatePresentationSubmission(presentation.presentation_submission)
+      }
+
+      ps.validatePresentation(presentationDefinition, presentation as unknown as VerifiablePresentation)
+      return true
+    } catch (e) {
+      agentContext.config.logger.error(e)
+      return false
+    }
+  }
+
+  public async getCredentialsForRequest(
+    agentContext: AgentContext,
+    { requestAttachment }: ProofFormatGetCredentialsForRequestOptions<PresentationExchangeProofFormat>
+  ): Promise<Array<W3cCredentialRecord>> {
+    const ps = this.presentationExchangeService(agentContext)
+    const { presentation_definition: presentationDefinition } = requestAttachment.getDataAsJson<{
+      presentation_definition: PresentationDefinition
+    }>()
+
+    ps.validatePresentationDefinition(presentationDefinition)
+
+    const presentationSubmission = await ps.selectCredentialsForRequest(agentContext, presentationDefinition)
+
+    const credentials = presentationSubmission.requirements.flatMap((r) =>
+      r.submissionEntry.flatMap((e) => e.verifiableCredentials)
+    )
+
+    return credentials
+  }
+
+  public async selectCredentialsForRequest(
+    agentContext: AgentContext,
+    { requestAttachment }: ProofFormatSelectCredentialsForRequestOptions<PresentationExchangeProofFormat>
+  ): Promise<Array<W3cCredentialRecord>> {
+    const ps = this.presentationExchangeService(agentContext)
+    const { presentation_definition: presentationDefinition } = requestAttachment.getDataAsJson<{
+      presentation_definition: PresentationDefinition
+    }>()
+
+    ps.validatePresentationDefinition(presentationDefinition)
+
+    const presentationSubmission = await ps.selectCredentialsForRequest(agentContext, presentationDefinition)
+
+    const credentials = presentationSubmission.requirements.flatMap((r) =>
+      r.submissionEntry.flatMap((e) => e.verifiableCredentials)
+    )
```

Comment:
As mentioned in other comments, the structures returned by the service adds a lot of needed context. I think we should return them as is, rather than flatmapping